cmd/backup/script.go

```diff
@@ -369,38 +436,92 @@ func (s *script) stopContainers() (func() error, error) {
 		}
 	}
 
-	var stopError error
-	if len(stopErrors) != 0 {
-		stopError = fmt.Errorf(
-			"stopContainers: %d error(s) stopping containers: %w",
-			len(stopErrors),
-			errors.Join(stopErrors...),
-		)
+	var scaledDownServices []swarm.Service
+	var scaleDownErrors []error
+	if isDockerSwarm {
+		for _, svc := range servicesToScaleDown {
+			service, _, err := s.cli.ServiceInspectWithRaw(context.Background(), svc.serviceID, types.ServiceInspectOptions{})
+			if err != nil {
+				scaleDownErrors = append(
+					scaleDownErrors,
+					fmt.Errorf("(*script).stopContainersAndServices: error inspecting service %s: %w", svc.serviceID, err),
+				)
+				continue
+			}
+			var zero uint64 = 0
+			serviceMode := &service.Spec.Mode
+			switch {
+			case serviceMode.Replicated != nil:
+				serviceMode.Replicated.Replicas = &zero
+			default:
+				scaleDownErrors = append(
+					scaleDownErrors,
+					fmt.Errorf("(*script).stopContainersAndServices: labeled service %s has to be in replicated mode", service.Spec.Name),
+				)
+				continue
+			}
+
+			response, err := s.cli.ServiceUpdate(context.Background(), service.ID, service.Version, service.Spec, types.ServiceUpdateOptions{})
+			if err != nil {
+				scaleDownErrors = append(scaleDownErrors, err)
+				continue
+			}
+
+			for _, warning := range response.Warnings {
+				s.logger.Warn(
+					fmt.Sprintf("The Docker API returned a warning when scaling down service %s: %s", service.Spec.Name, warning),
+				)
+			}
+
+			if err := progress.ServiceProgress(context.Background(), s.cli, service.ID, discardWriter); err != nil {
```

Comment:
It occurs to me this returns slightly earlier than containers have actually been removed, creating potential races between volume consumers (in my setup the app still holds the lock on a SQLite database, making the backup fail).

If I put a short sleep of a second after this line, it works, but obviously this shouldn't be the proper solution. Maybe this should instead check if there are no more containers that are children of this service or similar.

I'll need to look into this still.

Reply:
It seems adding

```go
			// progress.ServiceProgress returns too early, so we need to manually check
			// whether all containers belonging to the service have actually been removed
			for {
				containers, _ := s.cli.ContainerList(context.Background(), types.ContainerListOptions{
					Filters: filters.NewArgs(filters.KeyValuePair{
						Key:   "label",
						Value: fmt.Sprintf("com.docker.swarm.service.id=%s", service.ID),
					}),
				})
				if len(containers) == 0 {
					break
				}
				time.Sleep(time.Second)
			}
```
instead of sleeping actually works, although I am not sure if this is expected or if I am just using `ServiceProgress` incorrectly.

Reply:
The related command `docker service scale` will converge a bit before the actual containers are removed. Seems like the related API calls behave the same way.